bpf: Do not pre-allocate BPF maps by default

tgraf opened the pull request:
Set the daemon option to by default have the same setting as previous
versions of Cilium (enable preallocation where supported).
    
Update the kubernetes YAMLs such that users who upgrade just the
DaemonSet will have minimal upgrade impact as the current maps will be
retained with pre-allocated values. At the same time, add a default
option to the ConfigMap such that new users will automatically deploy
without pre-allocation, minimizing memory usage in newer deployments.
    
For more information on the new setting, see the ConfigMap comments.

Individual commit logs have further information.

Related: #6351 

<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/6357)
<!-- Reviewable:end -->

tgraf committed:
bpf: Be less aggressive when pre-allocating BPF maps

The default so far has been to pre-allocate all BPF maps. This has performance
benefits and avoids RCU free pile up but comes at a steep cost of memory
footprint.

Disable pre-allocation for all non-LRU map types and provide an option
--preallocate-bpf-maps to re-enable the behavior.

Signed-off-by: Thomas Graf <thomas@cilium.io>
Co-authored-by: Joe Stringer <joe@cilium.io>
Signed-off-by: Joe Stringer <joe@cilium.io>

borkmann commented on bpf/bpf_lxc.c:
```diff
@@ -62,6 +62,7 @@ struct bpf_elf_map __section_maps CT_MAP_TCP6 = {
 	.size_value	= sizeof(struct ct_entry),
 	.pinning	= PIN_GLOBAL_NS,
 	.max_elem	= CT_MAP_SIZE_TCP,
+	.flags		= DEFAULT_MAP_FLAGS,
```
LRU maps in non-prealloc mode cannot be created in the kernel.

tgraf replied:
Right, I was testing on 4.9 without LRU to get started. I've now made it conditional.

borkmann commented on bpf/lib/maps.h:
```diff
@@ -173,7 +179,7 @@ struct bpf_elf_map __section_maps cilium_ipcache = {
 	.pinning	= PIN_GLOBAL_NS,
 	.max_elem	= IPCACHE_MAP_SIZE,
-	.flags		= BPF_F_NO_PREALLOC,
+	.flags		= CONDITIONAL_PREALLOC,
```
This has to be fixed BPF_F_NO_PREALLOC as LPM map doesn't support prealloc.

borkmann reviewed (approved):
Aside from the one issue lgtm

joestringer committed:
daemon: Retain map preallocation settings by default

Set the daemon option to by default have the same setting as previous
versions of Cilium (enable preallocation where supported).

Update the kubernetes YAMLs such that users who upgrade just the
DaemonSet will have minimal upgrade impact as the current maps will be
retained with pre-allocated values. At the same time, add a default
option to the ConfigMap such that new users will automatically deploy
without pre-allocation, minimizing memory usage in newer deployments.

For more information on the new setting, see the ConfigMap comments.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
examples: Regenerate YAMLs

Regenerate YAMLs based on the recent changes.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
test: Enable BPF preallocation by default

Enable the upcoming BPF preallocation by default in the tests, so that
upgrade can be undertaken with minimal impact. Note that if this option
is toggled during upgrade, then it may have adverse effects.

This is being added because when it is set to false, the k8s upgrade
test is regularly failing, likely because the deletion of existing maps
causes etcd connectivity issues, which requires the etcd operator to
delete the pods and recreate them to get back into a good state. This
typically takes longer than the timeouts specified in the upgrade test,
causing the test to fail.

Further investigation is required to establish the exact behaviour for
toggling this option and how to safely enable/disable it; and also to
test the behaviour.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
daemon: Refactor service map open

Refactor this code into the loadbalancer.go file, which is a more
natural place for all of the service-related map open logic.

Signed-off-by: Joe Stringer <joe@cilium.io>

tgraf commented on examples/kubernetes/templates/v1/cilium-cm.yaml:
```diff
@@ -70,6 +70,21 @@ data:
 
+  # Pre-allocation of map entries allows per-packet latency to be reduced, at
+  # the expense of up-front memory allocation for the entries in the maps. The
+  # default below value will minimize memory usage in the default installation;
```
value below

tgraf commented on examples/kubernetes/templates/v1/cilium-cm.yaml:
```diff
@@ -70,6 +70,21 @@ data:
+  # connection for some time. Endpoints may need to be recreated to restore
+  # connectivity.
+  #
+  # For users upgrading from Cilium 1.3 or earlier, to minimize disruption
```
Let's rephrase this to say that Cilium <= 1.3 had this value to be always true so setting it to false and upgrading from <= 1.3 to >= 1.4 will cause some one-time disruptions during the upgrade. The way it is written right now, people may mistake it as for causing disruptions continuously.

Let's also specify that for versions < 1.4, this option has no effect. We should do this for other fields as well but doesn't hurt to start with this one.

joestringer replied:
@tgraf I adjusted for this feedback, PTAL.

ianvernon commented on daemon/daemon.go:
```diff
@@ -556,34 +556,8 @@ func (d *Daemon) init() error {
-			if _, err := proxymap.Proxy4Map.OpenOrCreate(); err != nil {
-				return err
-			}
+		if err := openServiceMaps(); err != nil {
```
Thanks for putting this into a separate function!

ianvernon commented on daemon/loadbalancer.go:
```diff
@@ -426,6 +427,40 @@ func (d *Daemon) RevNATDump() ([]loadbalancer.L3n4AddrID, error) {
+		if _, err := lbmap.RRSeq4Map.OpenOrCreate(); err != nil {
+			return err
+		}
+		if _, err := proxymap.Proxy4Map.OpenOrCreate(); err != nil {
```
Only slight nit I have here is that proxy maps are unrelated to the loadbalancer code, but I think having this code in a separate function is better than how it is currently in master.

ianvernon commented on examples/kubernetes/templates/v1/cilium-cm.yaml:
```diff
@@ -70,6 +70,24 @@ data:
+  # If this value is modified, then during the next Cilium startup the restore
+  # of existing endpoints and tracking of ongoing connections may be disrupted.
+  # This may lead to policy drops or a change in loadbalancing decisions for a
+  # connection for some time. Endpoints may need to be recreated to restore
```
This implies that the state is not able to be recovered over time? Or is it just that because the applications may get into a bad state themselves vs. in a bad state on the Cilium side?

joestringer replied:
~~Basically Cilium completely forgets about endpoints that were previously managed, and they must be restarted in order for the new instance of Cilium to start managing them.~~

EDIT: The above is wrong, that was with `clean-cilium-state` enabled in the ConfigMap.

In local upgrade testing of just the Cilium component, it surprisingly seems to update pretty seamlessly (admittedly without testing services, which I'd expect to break due to loadbalancing decisions being lost). I suspect that in the upgrade test, because the etcd instance is being redeployed, some more serious problems occur during bootstrap which take some time to get back into a good state.
